### ISSUE TITLE
Inconsistency with PointwiseMultiplication for SparseMatrix

### ISSUE THREAD

**Issue by ScottAlexanderCameron** [mathnet/mathnet-numerics#776]
In-place pointwise multiplication produces different results for `DenseMatrix` than it does for `SparseMatrix`.
example:
```c#
Matrix x = SparseMatrix.OfDiagonalArray(new double[] {1, 2, 3, 2});
Matrix y = DenseMatrix.OfDiagonalArray(new double[] {2, 1, 3, 2});
y.PointwiseMultiply(x, y);
Console.WriteLine(y);
```
produces the following output
```
DenseMatrix 4x4-Double
2  0  0  0
0  2  0  0
0  0  9  0
0  0  0  4
```

while for `SparseMatrix`, the output of the following code
```c#
Matrix x = SparseMatrix.OfDiagonalArray(new double[] {1, 2, 3, 2});
Matrix y = DenseMatrix.OfDiagonalArray(new double[] {2, 1, 3, 2});
x.PointwiseMultiply(y, x);
Console.WriteLine(x);
```
is
```
SparseMatrix 4x4-Double 0.00 % Filled
0  0  0  0
0  0  0  0
0  0  0  0
0  0  0  0
```

I have only tested this with matrices of doubles.
The root of this seems to be in the implementation of `DoPointwiseMultiply()` which calls `Clear()` on the result without checking whether or not the result matrix is one of the operands.

One would expect that the results are the same for `SparseMatrix` and `DenseMatrix` or at least an exception be thrown if this behaviour is not supported.

**Comment by jkalias** [comment]
Should be fixed in PR #912

**Comment by cdrnet** [comment]
Released in 5.0.0-beta2 - thank you both for reporting and for the patch!
